### DISCUSSION TITLE
Ophyd and Bluesky support in MXCube

### DISCUSSION THREAD

**Discussion by deepu9** [mxcube/mxcubecore#647] [General]
Thanks for all the hardwork behind MXCube software. I'm a Python developer from a synchrotron. We are using [Ophyd](https://github.com/bluesky/ophyd) and [Bluesky](https://github.com/bluesky/bluesky) at our synchrotron. When we are looking for a web UI, one of our scientists recommended MxCube. I did look into the code and played with it. I couldn't find a way to use Ophyd for devices instead of HardwareObject. Can you let me know If I can use Ophyd for devices in MXCube.

**Comment by beteva** [comment]
Dear @deepu9, I do not know Ophyd, but from what I could see on their github repository, it is what would correspond to [mxcubecore](https://github.com/mxcube/mxcubecore) for the MXCuBE GUI. Indeed in the MXCuBE GUI you address HardwareObjects, which are the abstraction layer for different type of devices and control systems, exactly like Ophyd does for Bluesky.
I see three solutions, if you want to use the MXCuBE GUI
- Step down one level and use Epics, which is already the case for several synchrotrons
- Check the signals and abstraction classes (e.g. AbstractActuator in mxcubecore would correspond to a certain level to positioner in Ophyd) and adapt your Ophyd code to mxcubecore.
- Look at the [mxcube3](https://github.com/mxcube/mxcube3) and adapt the python level to Ophyd.

**Comment by marcus-oscarsson** [comment]
Hi @deepu9 

I agree with @beteva but I would also add that It would probably be a good idea to have a quick zoom meeting so that we can discuss the details a bit better. Maybe you could explain how Ophyd is used from Bluesky and we can take it from there. Drop me an emial and we will set something up.

Marcus

**Comment by deepu9** [comment]
Thanks @beteva for your reply. I didn't quite understand your reply. From my knowledge Ophyd acts a hardware abstraction layer for EPICS and Bluesky is an orchestration layer which talks to EPICS and send details/instructions of Ophyd hardware objects. Happy to discuss more about the solutions you suggested.

Thanks @marcus-oscarsson . I'm happy to discuss through a zoom meeting. Couple of my colleagues are also willing to join the zoom meeting. Would it be ok with you?

**Reply by marcus-oscarsson** [reply]
Sure, lets try to setup something next week then just send me an email and we will try to manage.

**Reply by marcus-oscarsson** [reply]
@deepu9 you can send a mail to: oscarsso,at,esrf.fr

**Reply by beteva** [reply]
@deepu9, apparently my explanation was not clear enough, sorry. What I was saying is that Ophyd and mxcubecore have the same functionality. As @marcus-oscarsson suggests, the best would be to have a zoom meeting. Out of curiosity, which synchrotron are you working at.

**Reply by deepu9** [reply]
Apologies for late reply.

I work for Australian Nuclear Science and Technology Organisation(ANSTO). Me and my colleagues are excited to meet the makers of MXCube. At the moment we are exploring MXCube3 and MXCubeCore; and gaining as much knowledge as we can before we meet you. Can I come back to you with a proper date and time. Thanks

**Reply by marcus-oscarsson** [reply]
@deepu9 No problem, just let us know.
